### ISSUE TITLE
Enable lambdas to be created from ACK CRD

### ISSUE THREAD

**Issue by schollii** [aws/aws-controllers-k8s#197]
**Is your feature request related to a problem?**

Currently if I want serverless functionality to interact with a k8s cluster (due to how incredibly scalable lambdas are), I have to use a separate tool to create lambdas, roles etc (terraform, AWS SDK, serverless framework, are some of the main ones) and use either terraform or AWS SDK to create security groups so the lambdas can call back into the cluster. Yet lambdas are just another form of "container runner", just like Fargate and ECS. 

**Describe the solution you'd like**

It would be awesome to be able to describe a lambda in a manifest, pointing to a zip for the code, another for the layers, etc stored in s3 or (hey there's a thought!) in ECR (ECR now stores helm charts as artifacts, why not lambda zips and lambda layer zips!). It should be possible for a kubernetes manifest to indicate "I want a serverless function to run like this".

It would also be nice for the lambda to have a proxy in the cluster, eg to see its logs and query it and call it etc. This would make testing lambdas so easy, and perhaps eliminate the need for serverless framework (some ideas: the proxy could be just an API like  `kubectl lambda call/get...` type commands, which becomes possible with ACK since lambda becomes a new resource type, or could be a container that runs and communicates with the lambda, perhaps on demand and when shelled in you have 
 an high-level aws sdk to interact with the lambda lambda from within the cluster.

**Describe alternatives you've considered**

Terraform, AWS SDK, serverless

**Comment by jaypipes** [comment]
@schollii coming back to this today (sorry for the long delay!). We've made a bunch of progress in ACK since you first posted this issue. Recently, we even got AWS StepFunctions out into developer preview state:

https://github.com/aws/aws-controllers-k8s/issues/239

There is a GH issue for the ACK service controller for Lambda here:

https://github.com/aws/aws-controllers-k8s/issues/238

It is in the [`Planned` release status](https://aws.github.io/aws-controllers-k8s/releases/#planned).

One thing I'd like to note is that it is unlikely that ACK service controllers will ever manage data plane operations. For now, we're focusing on control plane operations, so Lambda APIs like [`Invoke`](https://docs.aws.amazon.com/lambda/latest/dg/API_Invoke.html) will probably not be something that ACK will deal with. Instead, control plane API calls like [`CreateFunction`](https://docs.aws.amazon.com/lambda/latest/dg/API_CreateFunction.html) and [`CreateEventSourceMapping`](https://docs.aws.amazon.com/lambda/latest/dg/API_CreateEventSourceMapping.html) would be the type of API calls that the ACK service controller for Lambda would be calling.

You can read more about my thoughts on the data plane calls in this ticket:

https://github.com/aws/aws-controllers-k8s/issues/292

Best,
-jay

**Comment by schollii** [comment]
Awesome to hear about #238.

Since ACK is focussed on provisioning it makes sense that operations like Invoke will not be covered. Would be nice to be able to completely configure the lambda including ref to layers stored in s3, triggers, tags, test events, etc.

BTW I suppose the spec will specify an s3 object path? Will controller be listening to events on that s3 object and its layers (which are also in s3), esp `updated` and `destroyed` events, and take appropriate action?

**Comment by jaypipes** [comment]
> Since ACK is focussed on provisioning it makes sense that operations like `Invoke` will not be covered. Would be nice to be able to completely configure the lambda including ref to layers stored in s3, triggers, tags, define test events, etc.

Yep. All configuration of the Lambda `Function`, including layers, tags, and event source mappings will be modifiable via ACK.

> BTW I suppose the spec will specify an s3 object path? Will controller be listening to events on that s3 object and its layers (which are also in s3), esp `updated` and `destroyed` events, and take appropriate action?

Good question. I suppose we could make an argument both for and against the ACK service controller doing this kind of event polling/listening. I think for at least the developer preview and beta releases of the Lambda ACK service controller we would *not* do this kind of a polling but perhaps later on we can write something for it.

**Comment by arashout** [comment]
Hey @jaypipes kind of a k8s newbie here and having trouble understanding the control-plane and data-plane terminology.     
Does calls like `UpdateFunctionCode` count as control plane operations? will they be included in the Lambda Controller functionality?    
e.g Could I update my lambda function with new code based on a zip in s3 using this controller?

**Comment by jaypipes** [comment]
> Hey @jaypipes kind of a k8s newbie here and having trouble understanding the control-plane and data-plane terminology.
> Does calls like `UpdateFunctionCode` count as control plane operations? will they be included in the Lambda Controller functionality?
> e.g Could I update my lambda function with new code based on a zip in s3 using this controller?

Hi @arashout! Sorry for the delayed response! We've been busy splitting our monorepo into smaller repos...

To answer your question, `UpdateFunctionCode` counts as a control plane operation, yes. You can consider CRUD+L (create, read, update, delete and list) operations on `Function` resources in Lambda to be control plane operations.

On the other hand, *executing* a particular Lambda function (as a result of a manual call or some event firing) would be an example of the Lambda *data plane*. Hope that clears things up! :)

**Comment by arashout** [comment]
Ah that makes sense. Thanks for the explanation, this would fit our usecase.     
I noticed that that this task related to control plane Lambda operations has been moved to `BUILD` in the project management chart.      
Is there an estimated timeline or is it too early to tell?
